### ISSUE TITLE
`Add all missing imports` Command/Code Action

### ISSUE THREAD

**Issue by OmarTawfik** [rust-lang/rust-analyzer#12761]
The current VS Code extension already offers code actions to add missing imports, but users have to iterate over all missing ones manually to add them. This effort is compounded when a large mod/function is copied/pasted into a different file, without the imports it uses. Users would manually have to:

1. Find the line with the diagnostic/error about the missing import.
3. Trigger code actions list.
4. Execute the `Add Foo` code action.
5. Possibly choose between multiple possible import sources.
6. Save the file.
7. Wait for updated diagnostics from `cargo check`.
8. Look for the next error.
9. Rinse and repeat for each one.

My suggestion is to offer a new command (and possibly a code action) titled `Add all missing imports`. Since the functionality is already there, I suspect it is just a matter of:

1. Iterate on all current "missing imports" diagnostics, and for each:
2. If import already added, skip it.
3. Add the missing import, and record it.
4. If there are multiple possible import sources, let the user choose one.

This way, the user workflow becomes:

1. Execute the `Add all missing imports` command.
7. If all missing imports have a single source, we are done!
8. If not, the choose dialogs can be shown in order, to select the ambiguous ones.

I suspect this is not perfect, and there can be corner cases, like importing two different symbols with the same name, but different paths. But for the amount of time this will save, I think most users would be ecstatic to quickly review the added imports afterwards to fix any issues. At least I would be :)

Feature already exists in other languages like TypeScript: [[Release Notes](https://github.com/microsoft/vscode-docs/blob/vnext/release-notes/v1_46.md#add-all-missing-imports-source-action)] [[Source Code](https://github.com/microsoft/TypeScript/blob/2ef3901940894064fd933a8de5b1c0d2a2262884/src/services/codefixes/importFixes.ts#L1035)]

![image](https://media.githubusercontent.com/media/microsoft/vscode-docs/8fcaefd8018572bcdecbc76f316a739fe522345a/release-notes/images/1_46/ts-import-all-source-action.gif)

Related to and possibly complements #5131 for improving the overall UX when working with large file imports.
Thanks for the amazing work on the extension!

**Comment by OmarTawfik** [comment]
Thanks for accepting the suggestion!

> We don't have a way of prompting the user when there are ambiguous imports

This is what I mean by prompting the user when there are ambiguities:

<img width="600" alt="image" src="https://user-images.githubusercontent.com/15987992/179860286-1e5b5f13-9302-4631-a587-e0c3da278d05.png">

**Comment by mankinskin** [comment]
I would also like this feature. I have started modifying the rust-analyzer cli to add a subcommand for applying all missing import fixes for a given file.

**Comment by osiewicz** [comment]
Couldn't we use [ShowMessageRequest](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#showMessageRequestParams) to prompt user in case of ambiguous imports?

**Comment by OmarTawfik** [comment]
> Couldn't we use [ShowMessageRequest](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#showMessageRequestParams) to prompt user in case of ambiguous imports?

@osiewicz If I can suggest, `showMessageRequest` is not document-specific, and lacks filtering, search, and other capabilities that are offered by the VS Code dialogs that Rust Analyzer already uses for imports ([example](https://github.com/rust-lang/rust-analyzer/issues/12761#issuecomment-1189620035)).

**Comment by osiewicz** [comment]
Yeah, that's fair, though VSC is not the only consumer of RA (albeit it's definitely the biggest one). I'd love to have `add all missing imports` in Zed, hence my suggested solution is not specific to VSC.

**Comment by OmarTawfik** [comment]
@osiewicz Good point. I looked more into it, and it looks like feature can actually be supported in an idiomatic way similar to how other languages already do it, but there is a tiny issue in Rust Analyzer that prevent this from happening: It needs to set the correct `kind` property on generated code actions. It currently just sets all of them to `quickfix`, even for missing imports, which is incorrect. It might consider switching to using the new scoped [CodeActionKind](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeActionKind) kinds and use a scoped string `source.addMissingImports `.

This allows users to set key bindings to add all missing imports in one step:

```jsonc
{
    "key": "ctrl+shift+i",
    "command": "editor.action.sourceAction",
    "args": {
        "kind": "source.addMissingImports",
        "apply": "first" // or "never" or "ifSingle"
    }
}
```

Or even do it on every file save:

```jsonc
"editor.codeActionsOnSave": [
    "source.addMissingImports"
]
```

More information in this release note: https://github.com/microsoft/vscode-docs/blob/vnext/release-notes/v1_46.md#add-all-missing-imports-source-action

And checking Zed docs, it seems the same method is also supported via `code_actions_on_format` and `codeActionOnSave` settings: https://zed.dev/docs/configuring-zed#code-actions-on-format

**Comment by Tuditi** [comment]
I would love to have this as well! Is there any existing solution for this?

**Comment by wyhinton** [comment]
Would love this feature as well!
